### ISSUE TITLE
I cannot understand where I can find the output

### ISSUE THREAD

**Issue by ucbtsh6** [rwanwork/sam2irs#1]
Hello 
I'm using this package However I don't know where can I find the output GTF and to understand it?

**Comment by rwanwork** [comment]
I think I need more information about what you have tried.  As a starting point, the repository includes an example.  Were you able to run it and get the output?  There is also an explanation about that file in the top-level `README.md` file.

Note that the output file is not a "GTF file" in the traditional sense.  However, we used its file format (since it's a popular file format) to indicate intron retention.

**Comment by ucbtsh6** [comment]
Hello,
Thanks for your reply. but I mean where is this out put file I cannot see any output file rather than print on the bash

**Comment by ucbtsh6** [comment]
#! /bin/bash
#$ -S /bin/bash
#$ -l tmem=30G
#$ -j y
#$ -N sam2.ir.52
#$ -o /SAN/transcriptome/
#$ -e /SAN/transcriptome/
#$ -wd /SAN/transcriptome/
export PATH=/share/apps/perl/bin:$PATH


cat /SAN//germany.fastq.files/52.fastq.sam |/SAN/transcriptome/sam2irs-master/Perl/sam2irs.pl --verbose --chrlist /SAN/transcriptome/sam2irs-master/Examples/test.sizes --gtf /SAN/transcriptome/REF/genome/transcripts.gtf

**Comment by rwanwork** [comment]
Oh!  I see what you mean.  So, the `sam2irs.pl` script sends the output to standard out.  If you ran it according to the example:

`cat test.sam | ../Perl/sam2irs.pl --verbose --chrlist test.sizes --gtf test1.gtf 2>/dev/null`

then the output would appear on your screen.  If you want it to a file, then you would have to send it to a file explictly:

`cat test.sam | ../Perl/sam2irs.pl --verbose --chrlist test.sizes --gtf test1.gtf 1>myfile.txt 2>/dev/null`

Then, the output would appear in `myfile.txt`.  As you're running it on a computing cluster, I guess it's appearing in the default output file for the system you're using (SGE?).  But you can just fix the problem by adding the `1>` part in.  You should also add the `2>` part as well, to remove any warning/error messages to standard error.

I should probably make this more explicit in the README -- thanks for the feedback!  I'll do that now, as well.

**Comment by ucbtsh6** [comment]
Hello 
Thanks you for your reply 
I will do that and see 
another question can I run all the samples parallel to get one output has the result of everything?

**Comment by ucbtsh6** [comment]
Another thing in my mind
I got this 
**[shassan@pchuckle shaimaa]$ cat  sam2.ir.44.o2431437 
II	Chromosome list:  /SAN/bahlerlab/killifish_transcriptome/shaimaa/sam2irs-master/Examples/test.sizes
II	GTF file:  /SAN/bahlerlab/killifish_transcriptome/shaimaa/IRFinder-1.3.0/REF/genome/transcripts.gtf
II	SAM record file:  Not provided
II	Output exons:  No
**
no SAM ?????
is there a problem?

**Comment by ucbtsh6** [comment]
I tried your new code it gives me empty  .txt file and nothing written in the file and on the output as well

**Comment by rwanwork** [comment]
No, you cannot run the samples in parallel.  You'll get one output GTF file for each sample.  How you combine them is not obvious to me.  i.e., is average best?  Or maximum?

As for the "SAM record file", perhaps I should have picked a better name...  It isn't referring to the input SAM file.  It is referring to a summary file that summarized various statistics.  You don't need it, but if you want to see it, then provide the `--samrecord` parameter with a filename.

As your original issue appears to have been solved, I'll close this issue.  If you have a separate issue later, feel free to open another issue with a "Subject" that reflects the issue you're facing.  Hopefully, you won't face many issues afterwards!

**Comment by rwanwork** [comment]
@ucbtsh6  Sorry, for the confusion, but I just clicked "close" on this issue just when you were typing your's out.  

Can you provide more information describing what you're providing as input?

Also, perhaps you can run it without the cluster first.  See if it works on the command-line?

**Comment by ucbtsh6** [comment]
the sam file
GTF
chrlist

cat /SAN/transcriptome/germany.fastq.files/ERR44_1.fastq.sam |/ SAN/transcriptome//sam2irs-master/Perl/sam2irs.pl --verbose --chrlist /SAN/transcriptome/sam2irs-master/Examples/test.sizes --gtf /SAN/transcriptome//transcripts.gtf 1>sam2.ir.44.txt 2>/dev/null 


1- chromosome list
scaffold02802	7250
scaffold02807	916
scaffold02808	6567
scaffold02809	9556
scaffold02810	4055
scaffold02811	249
scaffold02821	6921
scaffold02822	811
scaffold02829	705
scaffold02835	573
scaffold02844	5102
scaffold02854	1399
scaffold02859	4152
scaffold02861	814
scaffold02863	873
scaffold02865	5873
scaffold02876	1659
scaffold02882	4486
scaffold02889	3563
scaffold02892	3549
scaffold02893	6083
scaffold02894	637
scaffold02898	6379
scaffold02899	4345
scaffold02909	4389
scaffold02912	1840
scaffold02913	7867
scaffold02922	4330
scaffold02939	2508
scaffold02941	7567
scaffold02948	5539
scaffold02954	3950
2 GTF file
scaffold00001	nfurzeri_genebuild_v1	exon	10326	10341	.	-	.	transcript_id "Nfu_t_1_009268"; gene_id "Nfu_g_1_003772"; gene_name "CABZ01087670.1"; gene_biotype "protein_coding"; proteinid "Nfu_p_1_009268"; transcriptid "Nfu_t_1_009268"; annotation "DEAQ box RNA-dependent ATPase 1"; gene_biotype "protein_coding"; db_xref "ENSDARG00000059455%2CENSXMAG00000010692"; source "EVM_PASA";
scaffold00001	nfurzeri_genebuild_v1	CDS	10326	10341	.	-	1	transcript_id "Nfu_t_1_009268"; gene_id "Nfu_g_1_003772"; gene_name "CABZ01087670.1"; gene_biotype "protein_coding"; proteinid "Nfu_p_1_009268"; transcriptid "Nfu_t_1_009268"; annotation "DEAQ box RNA-dependent ATPase 1"; gene_biotype "protein_coding"; db_xref "ENSDARG00000059455%2CENSXMAG00000010692"; source "EVM_PASA";
scaffold00001	nfurzeri_genebuild_v1	exon	11449	11537	.	-	.	transcript_id "Nfu_t_1_009268"; gene_id "Nfu_g_1_003772"; gene_name "CABZ01087670.1"; gene_biotype "protein_coding"; proteinid "Nfu_p_1_009268"; transcriptid "Nfu_t_1_009268"; annotation "DEAQ box RNA-dependent ATPase 1"; gene_biotype "protein_coding"; db_xref "ENSDARG00000059455%2CENSXMAG00000010692"; source "EVM_PASA";
scaffold00001	nfurzeri_genebuild_v1	CDS	11449	11537	.	-	0	transcript_id "Nfu_t_1_009268"; gene_id "Nfu_g_1_003772"; gene_name "CABZ01087670.1"; gene_biotype "protein_coding"; proteinid "Nfu_p_1_009268"; transcriptid "Nfu_t_1_009268"; annotation "DEAQ box RNA-dependent ATPase 1"; gene_biotype "protein_coding"; db_xref "ENSDARG00000059455%2CENSXMAG00000010692"; source "EVM_PASA";
scaffold00001	nfurzeri_genebuild_v1	exon	12596	12639	.	-	.	transcript_id "Nfu_t_1_009268"; gene_id "Nfu_g_1_003772"; gene_name "CABZ01087670.1"; gene_biotype "protein_coding"; proteinid "Nfu_p_1_009268"; transcriptid "Nfu_t_1_009268"; annotation "DEAQ box RNA-dependent ATPase 1"; gene_biotype "protein_coding"; db_xref "ENSDARG00000059455%2CENSXMAG00000010692"; source "EVM_PASA";
3 sam file
ERR883844.7110643.1	256	sgr11	3437251	1	50M	*	0	0	CTGTTCCAGAGTTGTCAAGACAATAGTTCAATACCTATTTCATATTATTT	BB@4ADEFHHHFHJIIJJJJIJJJJJJJJJJIJJJJJJJJJIIGIJJJJI	AS:i:-3	ZS:i:-3	XN:i:0	XM:i:1	XO:i:0	XG:i:0	NM:i:1	MD:Z:3G46	YT:Z:UU	NH:i:5
ERR883844.7110644.1	0	sgr15	8300226	60	50M	*	0	0	CTTTGGTGTGTCTCGTGCTTTGGGCGTACTCGCTCAGCTGGTGTGGAGTA	=<@+=A=BDFDFHIHHIIGIIIGB9EEEDEGHIIII@CHHGBB8BF;C@G	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:50	YT:Z:UU	NH:i:1
ERR883844.7110645.1	16	scaffold01456	19351	60	50M	*	0	0	GAGTTGTATTTATCAGGGGGGGTGCGCCCATCAAACACCCTCCATCTCCA	CDEEFFFFFFHHHHHJJIJJJJJIGIJJIJJJIHAGFHGHHHFFFDDCCB	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:50	YT:Z:UUNH:i:1
ERR883844.7110646.1	16	scaffold00008	331182	60	50M	*	0	0	TGGCGCGAAAGATGAAGGATACGGATAGCGAGGAAGAAATCAGGGAGGCG	HIGHGGCFIHCE@JGGHHG;IEIGGJIGGJJIIGGJIHHFHHFFDA+@@@	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:50	YT:Z:UUNH:i:1
ERR883844.7110647.1	16	sgr01	95543765	60	50M	*	0	0	GTGGTGCGGGAGATGTTCAGCAGGTCCTCCTGGAAACGAGCCGGGTAGTT	DCADA<@;=BDDBB@??@DB3D=DCD=DE>EBEE;EEDDDD0BD?A4?8?	AS:i:-3	ZS:i:-4	XN:i:0	XM:i:1	XO:i:0	XG:i:0	NM:i:1	MD:Z:46C3	YT:Z:UU	NH:i:1
ERR883844.7110648.1	4	*	0	0	*	*	0	0	TTCAGTGCATTTTGGGGGGAATCTTTTTTTTTTGTTTATTTGTTTTCATG	CCCFFFFFHHHHHJJJJJDDDDDDDDDDDDDDDBBDD>CDDDDDDDDDEA	YT:Z:UU
ERR883844.7110649.1	0	sgr01	30918125	60	50M	*	0	0	CGCCTTCGCCAGTGTCATCTTGTTTGGCACCAACAATGACAGCAGCATCT	==7+=@DDFFFFFD<CEFIEHFDHFFCEFFE@F=DE<FFI<?;B?BF?FF	AS:i:0	XN:i:0	XM:i:0	XO:i:0	XG:i:0	NM:i:0	MD:Z:50	YT:Z:UUNH:i:1

**Comment by ucbtsh6** [comment]
it gives output print on screen if I didn't provide 1> & 2> 
but it gives empty files if I specify them

**Comment by ucbtsh6** [comment]
this is the output in the bash if I don't specify 1> & 2> I Size of chromosome:  168361
II   Intron positions:  120291 (71.45%)
II   Exon positions:  5256 (3.12%)
II   Non-intron/exon positions:  42815 (25.43%)
S4 scaffold00097
168361 120291
5256 42815
II Total bases:  26198
II SAM entries:
II   Unavailable CIGAR string (skipped):  0
II   With N (skipped):  65192
II   Without N (used):  268303
S5 scaffold00097
26198 0
65192 268303
II Intron retention genes:  3
II Intron retentions:  29
II Non-intron retentions:  20
S6 scaffold00097
3 29
20 
II Total number of genes:  9
II   Overlapping:  4
II   Not overlapping:  5
II List of genes:
GG   Nfu_g_1_017966
GG   Nfu_g_1_017967
GG   Nfu_g_1_017970
GG   Nfu_g_1_017971
S3 scaffold00098
9 4
5 
II Size of chromosome:  178595
II   Intron positions:  26670 (14.93%)
II   Exon positions:  4509 (2.52%)
II   Non-intron/exon positions:  147417 (82.54%)
S4 scaffold00098
178595 26670
4509 147417
how to read it?

**Comment by rwanwork** [comment]
Thank you for your patience and the feedback!  Indeed, this is good information for me as I can try to make the software (or its documentation) better.  I'm sorry that you have to go through this.

A quick look of your input files seems fine.

As for the output when you remove the "2>", this is actually debugging output.  It does tell me that the program is running.  The S1, S2, S3, S4, S5, S6, and S7 were used by me to indicate different steps through the program.  I take it what you're showing me is just part of the entire output?  I think the top and the bottom would be interesting.

S1 - How it read your GTF file
S2 - How it read your SAM file
S3 to S6 - Its progress through each scaffold; there will be many of these.  We can ignore them for now.
S7 - Final summary

If it's not too much trouble, can you show me the S1, S2, and S7 sections?  That will tell me if it at least read your input fine and if it output something.  Even in that snippet above, I see the line "Intron retentions:  29"  So, on scaffold00097, it found 29 intron retentions.  So, yes, _something_ should have been printed and I'm curious what is wrong.

Sorry that you have to go through this, but I haven't seen this problem before.  So, you'll have to walk me through it.  If there is information that you can show me but prefer not to post publicly, you can e-mail me, if you want.

**Comment by ucbtsh6** [comment]
Hi Rwan,
Thanks a lot for your help
It doesn’t have any S7 yet this is not in the txt file it is when I don’t specify any out put so printed on the bash
What can I provide to help you
Regards
Shaimaa
[cid:46E6B44D-8EFA-4E91-8978-4A9279CF1E1A]


On 8 Oct 2020, at 17:35, Raymond Wan <notifications@github.com<mailto:notifications@github.com>> wrote:



Thank you for your patience and the feedback! Indeed, this is good information for me as I can try to make the software (or its documentation) better. I'm sorry that you have to go through this.

A quick look of your input files seems fine.

As for the output when you remove the "2>", this is actually debugging output. It does tell me that the program is running. The S1, S2, S3, S4, S5, S6, and S7 were used by me to indicate different steps through the program. I take it what you're showing me is just part of the entire output? I think the top and the bottom would be interesting.

S1 - How it read your GTF file
S2 - How it read your SAM file
S3 to S6 - Its progress through each scaffold; there will be many of these. We can ignore them for now.
S7 - Final summary

If it's not too much trouble, can you show me the S1, S2, and S7 sections? That will tell me if it at least read your input fine and if it output something. Even in that snippet above, I see the line "Intron retentions: 29" So, on scaffold00097, it found 29 intron retentions. So, yes, something should have been printed and I'm curious what is wrong.

Sorry that you have to go through this, but I haven't seen this problem before. So, you'll have to walk me through it. If there is information that you can show me but prefer not to post publicly, you can e-mail me, if you want.

**Comment by rwanwork** [comment]
@ucbtsh6 So, if "S7" isn't shown, then that tells me there was a problem with the execution of the program.  That would explain why you do not see any output.  (Output is provided at the very end, not while it is running.)

Sorry about the inconvenience!  I do appreciate your patience in helping me find out what is the cause.  I guess there is something in your input files which is not wrong, but most likely I did not anticipate.

Can you run your script like this:

    cat /SAN/transcriptome/germany.fastq.files/ERR44_1.fastq.sam | /SAN/transcriptome//sam2irs-master/Perl/sam2irs.pl --verbose 1 --chrlist /SAN/transcriptome/sam2irs-master/Examples/test.sizes --gtf /SAN/transcriptome//transcripts.gtf 1>sam2.ir.44.txt 2>error.log

I've basically changed the `--verbose` option so that there are levels.  I also made the output a bit more readable! 
 Printing everything might help me, but I think it would overwhelm both of us.  So, let's start with `1`.  You can either attach `error.log`, or if there are reasons you do not want to share it publicly, you can e-mail me.  At the very end should be the reason why it stopped.  That's what I'm looking for, but everything before it will help.

Thank you!

**Comment by ucbtsh6** [comment]
Hello Rwan
I did what you told me but yet the txt file is empty.
Is there anything else to do?

On 9 Oct 2020, at 03:45, Raymond Wan <notifications@github.com<mailto:notifications@github.com>> wrote:



@ucbtsh6<https://github.com/ucbtsh6> So, if "S7" isn't shown, then that tells me there was a problem with the execution of the program. That would explain why you do not see any output. (Output is provided at the very end, not while it is running.)

Sorry about the inconvenience! I do appreciate your patience in helping me find out what is the cause. I guess there is something in your input files which is not wrong, but most likely I did not anticipate.

Can you run your script like this:

cat /SAN/transcriptome/germany.fastq.files/ERR44_1.fastq.sam | /SAN/transcriptome//sam2irs-master/Perl/sam2irs.pl --verbose 1 --chrlist /SAN/transcriptome/sam2irs-master/Examples/test.sizes --gtf /SAN/transcriptome//transcripts.gtf 1>sam2.ir<http://sam2.ir>.44.txt 2>error.log


I've basically changed the --verbose option so that there are levels. I also made the output a bit more readable!
Printing everything might help me, but I think it would overwhelm both of us. So, let's start with 1. You can either attach error.log, or if there are reasons you do not want to share it publicly, you can e-mail me. At the very end should be the reason why it stopped. That's what I'm looking for, but everything before it will help.

Thank you!

**Comment by rwanwork** [comment]
@ucbtsh6 The `sam2.ir.44.txt` should be empty since the program did not complete correctly.  But not `error.log`.  It's `error.log` that I'm interested in...

**Comment by ucbtsh6** [comment]
Hello
It's empty as well
________________________________

**Comment by rwanwork** [comment]
@ucbtsh6 This is with the updated version of the software?

How about if you run it with the sample data?  Surely that should have something in both the file after `1>` and the file after `2>`?

Can you also try a higher verbose level, then?  Like with `--verbose 3`?

**Comment by rwanwork** [comment]
If there are no privacy issues, you can send your input files to me via e-mail.  If there are privacy issues, you can rename samples, etc. before sending them to me.

If there is no output being generated, then it is really difficult for me to understand what's going on without trying it myself...

**Comment by ucbtsh6** [comment]
Hello,
I think sending files are a good idea
GTF
https://drive.google.com/file/d/1GmAMBtrPVmf1i-AkEZ4VVSnltuYwNqpC/view?usp=sharing
SAM
https://drive.google.com/file/d/1tUoIchvbqxhyQ3Jhn2adStfaFiPT3SVQ/view?usp=sharing
Chromosome list
https://drive.google.com/file/d/1WeNFRzbbukrxlaW05gNkC6Ox3LTau2ej/view?usp=sharing

On 9 Oct 2020, at 10:34, Raymond Wan <notifications@github.com> wrote:



If there are no privacy issues, you can send your input files to me via e-mail. If there are privacy issues, you can rename samples, etc. before sending them to me.

If there is no output being generated, then it is really difficult for me to understand what's going on without trying it myself...

**Comment by ucbtsh6** [comment]
Hello

I think sending files is good idea
GTF

https://drive.google.com/file/d/1GmAMBtrPVmf1i-AkEZ4VVSnltuYwNqpC/view?usp=sharing

SAM

https://drive.google.com/file/d/1tUoIchvbqxhyQ3Jhn2adStfaFiPT3SVQ/view?usp=sharing

Chromosome list
https://drive.google.com/file/d/1WeNFRzbbukrxlaW05gNkC6Ox3LTau2ej/view?usp=sharing

On 9 Oct 2020, at 10:34, Raymond Wan <notifications@github.com<mailto:notifications@github.com>> wrote:



If there are no privacy issues, you can send your input files to me via e-mail. If there are privacy issues, you can rename samples, etc. before sending them to me.

If there is no output being generated, then it is really difficult for me to understand what's going on without trying it myself...

**Comment by rwanwork** [comment]
@ucbtsh6 

Thanks a lot for providing the files!  You may delete them from your Google Drive.  Just wanted to tell you that the SAM file has an error -- the last line is incomplete since only the read ID is present.  (You can see it by typing `tail -n 2 ER71.sam`.)  But you gave me enough and I know the file is very large.  

Indeed, I have been able to reproduce the problem you were seeing.  Please give me a few days to try to solve your problem.  Thanks again for your patience in helping me track down the cause of your problems!

**Comment by ucbtsh6** [comment]
Hello Rwan,
I’m not quite sure I understood you well. Do you mean that there is something wrong with my sam file?
Thats weird as I used it for feature count and worked pretty well?
This is the link of it
https://drive.google.com/file/d/174exirYyBCz3m1mrFYFZwXtGkYDazPlu/view?usp=sharing
Regards

On 12 Oct 2020, at 12:06, Raymond Wan <notifications@github.com<mailto:notifications@github.com>> wrote:

 tail -n 2 ER71.sam.

**Comment by rwanwork** [comment]
@ucbtsh6 Yes, there is something wrong with your file.  Like I suggested, you will see it if you type `cat ERR71.sam | tail -n 1`.  What you will see is this:

    ERR883837.17105714.1

The last line just has the name of a read, but nothing after.  As the SAM [standard](https://samtools.github.io/hts-specs/SAMv1.pdf) says in Section 1.4, "Each line consists of 11 or more TAB-separated fields. The first eleven fields are always present and in the order shown below; ".  So, this last line in your file is incomplete.

If you do not believe me still, you can try compressing it with `samtools`.  For example, try running this:

    $ samtools view -S -b ERR71.sam >ERR71.bam
    [W::sam_read1] Parse error at line 18994190
    [main_samview] truncated file.

Of course, I have *no* intention of shifting the blame on you.  My program does seem to have a problem and your file will help me try to figure out what's wrong.  (Please give me some time to sort it out.)  So, thank you!  

My comment was just an "FYI" to let you know that your file doesn't quite follow the standard.  Seeing it is good for me as I need to adapt to this kind of input, whether or not it follows the standard.  Thank you!

**Comment by ucbtsh6** [comment]
Hi Rwan,
I’m really sorry for much disturbance I made for you. I’m just a beginner in bioinformatics so I’m really not good in this details, so thanks a lot for your explanation. And I’m waiting for you. Tell me if anything I can do?
Thanks a lot again.
Regards,
Shaimaa


On 13 Oct 2020, at 16:10, Raymond Wan <notifications@github.com<mailto:notifications@github.com>> wrote:



@ucbtsh6<https://github.com/ucbtsh6> Yes, there is something wrong with your file. Like I suggested, you will see it if you type cat ERR71.sam | tail -n 1. What you will see is this:

ERR883837.17105714.1


The last line just has the name of a read, but nothing after. As the SAM standard<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fsamtools.github.io%2Fhts-specs%2FSAMv1.pdf&data=02%7C01%7C%7C85c56af14e02490cb89808d86f8a1f7a%7C1faf88fea9984c5b93c9210a11d9a5c2%7C0%7C1%7C637381986303836505&sdata=TuCvLXfC35VdPP%2B81c17N63x1avvQl9N68AzWh7C%2BX0%3D&reserved=0> says in Section 1.4, "Each line consists of 11 or more TAB-separated fields. The first eleven fields are always present and in the order shown below; ". So, this last line in your file is incomplete.

If you do not believe me still, you can try compressing it with samtools. For example, try running this:

$ samtools view -S -b ERR71.sam >ERR71.bam
[W::sam_read1] Parse error at line 18994190
[main_samview] truncated file.


Of course, I have no intention of shifting the blame on you. My program does seem to have a problem and your file will help me try to figure out what's wrong. (Please give me some time to sort it out.) So, thank you!

My comment was just an "FYI" to let you know that your file doesn't quite follow the standard. Seeing it is good for me as I need to adapt to this kind of input, whether or not it follows the standard. Thank you!

**Comment by ucbtsh6** [comment]
Hi Rwan,
I’m really sorry for much disturbance I made for you. I’m just a beginner in bioinformatics so I’m really not good in this details, so thanks a lot for your explanation. And I’m waiting for you. Thanks a lot again
Regards,
Shaimaa


On 13 Oct 2020, at 16:10, Raymond Wan <notifications@github.com<mailto:notifications@github.com>> wrote:



@ucbtsh6<https://github.com/ucbtsh6> Yes, there is something wrong with your file. Like I suggested, you will see it if you type cat ERR71.sam | tail -n 1. What you will see is this:

ERR883837.17105714.1


The last line just has the name of a read, but nothing after. As the SAM standard<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fsamtools.github.io%2Fhts-specs%2FSAMv1.pdf&data=02%7C01%7C%7C85c56af14e02490cb89808d86f8a1f7a%7C1faf88fea9984c5b93c9210a11d9a5c2%7C0%7C1%7C637381986303836505&sdata=TuCvLXfC35VdPP%2B81c17N63x1avvQl9N68AzWh7C%2BX0%3D&reserved=0> says in Section 1.4, "Each line consists of 11 or more TAB-separated fields. The first eleven fields are always present and in the order shown below; ". So, this last line in your file is incomplete.

If you do not believe me still, you can try compressing it with samtools. For example, try running this:

$ samtools view -S -b ERR71.sam >ERR71.bam
[W::sam_read1] Parse error at line 18994190
[main_samview] truncated file.


Of course, I have no intention of shifting the blame on you. My program does seem to have a problem and your file will help me try to figure out what's wrong. (Please give me some time to sort it out.) So, thank you!

My comment was just an "FYI" to let you know that your file doesn't quite follow the standard. Seeing it is good for me as I need to adapt to this kind of input, whether or not it follows the standard. Thank you!

**Comment by rwanwork** [comment]
@ucbtsh6 No worries!  Everyone comes into bioinformatics from a different background and with different prior experience.

I appreciate the time you've taken to explain to me your problems.  Indeed, the software cannot improve if I don't receive feedback from other users.

I'm afraid I have a busy week and am juggling a couple of unexpected things.  But I'll let you know as soon as I can get back to you.  Thank you again for your patience!

**Comment by rwanwork** [comment]
@ucbtsh6 So, I figured out one of the causes of the problems we are seeing.  It seems your input file is incorrect.

For example, `chromosome.list.txt` is supposed to list every "chromosome" in your input, as well as its length.  You have indicated that *scaffold00001* is 510,121 base pairs in length.  (First line of the file.)

However, in your GTF file, I can type this:  `grep scaffold00001 GTF.gtf  | head -n 4 | cut -f 1-5` and obtain this output:

    scaffold00001	nfurzeri_genebuild_v1	exon	516280	516309
    scaffold00001	nfurzeri_genebuild_v1	CDS	516280	516309
    scaffold00001	nfurzeri_genebuild_v1	exon	519566	520447
    scaffold00001	nfurzeri_genebuild_v1	CDS	519566	520447

(The command says to look for the phrase "scaffold00001" in the file, taking the first 4 lines and then the first 5 columns.)  As you can see, there is a gene with two exons from 516280 to 520447.

Moreover, I also see in your SAM file, you have a read called ERR883837.295962.1 that aligns to position 584323.

That said, I have to admit I could improve my code a bit better and give a more friendly error than silently failing.  I could probably throw away the requirement for "chromosome.list.txt".  So, I think I can do something for you in the next few days (I hope!) to make it easier for you and other future users.  Your problem has given me a few ideas on what I could have done. 
 
Please give me a bit more time, but you should look into how you're generating these files.  "chromosome.list.txt" doesn't match your other two files.

**Comment by ucbtsh6** [comment]
Hi Rwan,

I’m really thankful for all your trials and help and sorry for that mistake really
I will re do this file again and try.
For improvement purpose —> could you do something like featureCounts that I can run all the samples together and the output comes in one file
Regards

On 18 Oct 2020, at 11:18, Raymond Wan <notifications@github.com<mailto:notifications@github.com>> wrote:

ou can see, there is a gene with two exons from 51628

**Comment by ucbtsh6** [comment]
Hi Rwan
I re-made the chromosome list by other way but It still not working
This is the new size folder
I’m really sorry for disturbance


On 18 Oct 2020, at 11:18, Raymond Wan <notifications@github.com<mailto:notifications@github.com>> wrote:



@ucbtsh6<https://github.com/ucbtsh6> So, I figured out one of the causes of the problems we are seeing. It seems your input file is incorrect.

For example, chromosome.list.txt is supposed to list every "chromosome" in your input, as well as its length. You have indicated that scaffold00001 is 510,121 base pairs in length. (First line of the file.)

However, in your GTF file, I can type this: grep scaffold00001 GTF.gtf | head -n 4 | cut -f 1-5 and obtain this output:

scaffold00001   nfurzeri_genebuild_v1   exon    516280  516309
scaffold00001   nfurzeri_genebuild_v1   CDS     516280  516309
scaffold00001   nfurzeri_genebuild_v1   exon    519566  520447
scaffold00001   nfurzeri_genebuild_v1   CDS     519566  520447


(The command says to look for the phrase "scaffold00001" in the file, taking the first 4 lines and then the first 5 columns.) As you can see, there is a gene with two exons from 516280 to 520447.

Moreover, I also see in your SAM file, you have a read called ERR883837.295962.1 that aligns to position 584323.

That said, I have to admit I could improve my code a bit better and give a more friendly error than silently failing. I could probably throw away the requirement for "chromosome.list.txt". So, I think I can do something for you in the next few days (I hope!) to make it easier for you and other future users. Your problem has given me a few ideas on what I could have done.

Please give me a bit more time, but you should look into how you're generating these files. "chromosome.list.txt" doesn't match your other two files.

**Comment by rwanwork** [comment]
@ucbtsh6 I think there are other issues with my code.  I'm still using your input files to try to figure out what's wrong.  Let me get back to you in a few days (I hope).

I have updated the source code today.  Your problem isn't resolved, but `chromosome.list.txt` is no longer needed.  So, don't worry about that file any more.

As for your suggestion, I'm afraid how intron retention is calculated doesn't work like that.  Intron retention is calculated on a per-sample basis.  I suppose you can calculate it on each sample, and then combine them into a table.  But this would only be fore aesthetic reasons.  Running it across all files simultaneously would make the whole process more inefficient since each sample's intron retention is determined independently of all other samples.

**Comment by rwanwork** [comment]
@ucbtsh6 

I can't say that it is bug free but I believe I found the cause of the problem.  You can try the latest version from the repository.

From what I can tell, the cause of the problem was that your file of chromosome lengths was incorrect.  I then removed it a few days ago, and I made the program worse...

So, I need to go back on my "word" and reinstate the `chromosome.list.txt` file.  Now, when it reads the file and the chromosome length is "too short", it will tell you.  But it can only tell you that it is wrong.  It can't tell you what it should be.

How did you generate your original `chromosome.list.txt`?  It seems many entries are incorrect..  Since you aligned your reads to some kind of reference genome, it's probably better to get the length from the reference genome size.

Before you jump into running it on your cluster, try running it like this (fill in the correct paths for the filenames):

    cat ER71.sam | sam2irs.pl --chrlist chromosome.list.txt --verbose 2 --gtf GTF.gtf --log log.txt 1>out.txt 2>err.txt

You should get something in the 3 output files:  `out.txt`, `log.txt`, and `err.txt`.  I've tried my best to make the program a bit more informative.  Hopefully, this time around, your problem is either solved, or that it can tell you what is wrong.
